### DISCUSSION TITLE
cache-control header missing in Signed URL provided by Supabase storage

### DISCUSSION THREAD

**Discussion by mohammed-zia** [supabase/supabase#21926] [Questions] [answered]
Hi all. 

When I upload an image to a storage bucket, I set the `cache-control` header to 3600, and this is reflected when I check the image's metadata:

```
{'created_at': '2024-03-10T09:10:32.135Z',
  'id': '2d43437a-cdbd-4f2b-8e41-9b311640364e',
  'last_accessed_at': '2024-03-10T09:10:32.135Z',
  'metadata': {'cacheControl': '3600',
               'contentLength': 12933264,
               'eTag': '"a8293ac4e7cb9a8b0582601964382b6c"',
               'httpStatusCode': 200,
               'lastModified': '2024-03-10T09:10:32.045Z',
               'mimetype': 'image/image',
               'size': 12933264},
  'name': 'Screenshot 2024-02-14 at 15.26.58 (2).png',
  'updated_at': '2024-03-10T09:10:32.135Z'}
```

If I request the image via `curl -X GET -I 'http://localhost:8000/storage/v1/object/authenticated/...` I can see the `cache-control` header in the response:

```
HTTP/1.1 200 OK
Content-Type: image/image
Content-Length: 12933264
Connection: keep-alive
accept-ranges: bytes
etag: a8293ac4e7cb9a8b0582601964382b6c
last-modified: Sun, 10 Mar 2024 09:10:32 GMT
cache-control: 3600
Date: Sun, 10 Mar 2024 16:30:52 GMT
Access-Control-Allow-Origin: *
X-Kong-Upstream-Latency: 45
X-Kong-Proxy-Latency: 0
Via: kong/2.8.1
```

But if I generate a signed URL and then request the image via the browser, it doesn't include the cache-control header:

```
HTTP/1.1 200 OK
Content-Type: image/image
Content-Length: 12933264
Connection: keep-alive
accept-ranges: bytes
etag: a8293ac4e7cb9a8b0582601964382b6c
last-modified: Sun, 10 Mar 2024 09:10:32 GMT
expires: Sun, 10 Mar 2024 17:23:36 GMT
Date: Sun, 10 Mar 2024 16:23:36 GMT
Access-Control-Allow-Origin: *
X-Kong-Upstream-Latency: 51
X-Kong-Proxy-Latency: 1
Via: kong/2.8.1
```

I'm not sure if this is by design - perhaps I'm misunderstanding how Supabase storage is supposed to be used. Is the `cache-control` header supposed to be returned with the pre-signed URL?

For context, I'm self-hosting Supabase locally and interacting with it entirely via my backend with REST API requests. 

At the moment the images are not being cached in the client's browser when using the signed URLs, they are being re-fetched each time the page loads. I want to avoid this as the images are not likely to change often so it seems a waste of resources and time to re-fetch.

Any info appreciated, thanks in advance.

**Comment by GaryAustin1** [comment]
I believe expires header is taking the place of cache-control header.

A random hit from the web:
https://imagekit.io/blog/ultimate-guide-to-http-caching-for-static-assets/
![image](https://github.com/supabase/supabase/assets/54564956/3387cc52-8bf0-440c-8464-ff8dffe2d036)

I don't know the details of image caching but maybe pursue expires header and how that would work for your environment.

**Reply by mohammed-zia** [reply, marked as answer]
Hmm it does seem that `expires` _should_ handle caching, but for some reason it isn't working for me (tried both Firefox 123.0.1 and Safari v17.1).

In case anyone comes across this with the same issue - I've fixed it by proxying the image through my backend and explicitly setting the `cache-control` header there. Not the most efficient/scalable but it works for now.

Thanks anyway, loving Supabase!
